### PR TITLE
sstables: don't rewrite Statistics.db of an unlinked sstable

### PR DESCRIPTION
**Pull request by tarasveretilnyk**
update_repaired_at() -> rewrite_statistics() writes Statistics.db.tmp and renames it over Statistics.db without checking whether the sstable was unlinked meanwhile.  If anything deletes an sstable that an in-flight incremental repair has captured, before the repair marks it as repaired, the rename recreates a lone Statistics.db in a directory whose other components are already gone.  Nothing fails at the time.  The node refuses to boot on the next restart, which may be weeks later and unrelated:

    sstables::malformed_sstable_exception (At directory: ...: no TOC found
    for SSTable .../me-<generation>-big-Statistics.db!. Refusing to boot)

One reproducible path: an incremental_mode=full repair captures sstables that are already repaired, and being_repaired does not move them out of the repaired view, since the classifier tests is_repaired() before is_being_repaired().  The compaction disable and the incremental-repair write lock are taken only on view_for_unrepaired_data(), leaving the repaired view unprotected, so a compaction of it -- for instance the one ALTER TABLE triggers on every view -- can delete the captured sstables before the mark.

A second: the coordinator can move the tablet to end_repair while the repair master is still running -- in practice when the tablet_repair RPC to the master fails but the master carries on.  If the drain RPC to the master fails too, the coordinator only logs "drain rpc failed, proceed to fence old writes" and sends repair_update_compaction_ctrl, which clears being_repaired and releases the repair lock, so compaction on the master can delete the captured sstables before it marks them.  Both paths were reproduced: the first leaves orphans on every replica, the second on the repair master only.

Skip the rewrite when the sstable is already unlinked.  Nothing is lost by skipping: the deleted sstable's data lives on in the compaction output, which inherited repaired_at from its inputs, and is_repaired() accepts any non-zero repaired_at at or below the watermark, so that output stays classified as repaired.

The check is race-free because rewrite_statistics() takes _mutate_sem and unlink() sets _unlinked under that same semaphore; it has to be placed after the lock is acquired, not before.  Guarding here rather than in update_repaired_at() also covers the mutate_sstable_level() caller.

Master fixed this as a side effect of 5bbc44ed12 ("sstables: replace rewrite_statistics with new rewrite component mechanism").

Fixes https://scylladb.atlassian.net/browse/SCYLLADB-4582

Backport is required to 2026.1

### CODE REVIEW

**Review by denesb — APPROVED** [review]
@asias please review the two races indentified while this was investigated.
The one with full repair is concerning. Should we open issues for these (and fix them) or are they benign?

**Comment by raphaelsc** [comment]
I think this problem should be fixed systematically, rather than ignoring some mutation because the sstable has already been unlinked. A sstable that has been unlinked shouldn't be a candidate for mutation of statistics. I am not a fan of checking sstable has been unlinked. We have done so in this procedure but it may happen in others too.

A sstable in the set cannot be unlinked. A repairing sstable shouldn't be eligible for compaction, so what unlinked it before it was promoted to repaired set which happens by mutating repaired_at stat?

**Comment by raphaelsc** [comment]
This sounds like a bandaid to me, we should understand the root cause and fix it.

**Comment by denesb** [comment]
> This sounds like a bandaid to me, we should understand the root cause and fix it.

I think this fix is fine. This is an LTS version, so we should deliver surgical fixes.
There are two possible root-causes mentioned in the cover-letter. I asked @asias to look into them. We can pursue them in parallel.

**Comment by raphaelsc** [comment]
> > This sounds like a bandaid to me, we should understand the root cause and fix it.
> 
> I think this fix is fine. This is an LTS version, so we should deliver surgical fixes. There are two possible root-causes mentioned in the cover-letter. I asked @asias to look into them. We can pursue them in parallel.

I am not sure it is. The full repair might lose some compacted sstables in repaired set, those sstables will be demoted to unrepaired set (since they will escape the marking as repaired by mutating repaired_at) and excluded from tombstone gc.

**Comment by denesb** [comment]
> > > This sounds like a bandaid to me, we should understand the root cause and fix it.
> > 
> > 
> > I think this fix is fine. This is an LTS version, so we should deliver surgical fixes. There are two possible root-causes mentioned in the cover-letter. I asked @asias to look into them. We can pursue them in parallel.
> 
> I am not sure it is. The full repair might lose some compacted sstables in repaired set, those sstables will be demoted to unrepaired set (since they will escape the marking as repaired by mutating repaired_at) and excluded from tombstone gc.

Right, an sstable demoted from repaired -> unrepaired will mean it will be excluded from tombstone-gc overlap checks and can cause data resurrection. So we definitely have to look into the problem of compaction allowed on repaired set in parallel to full repair.

**Comment by denesb** [comment]
> > > > This sounds like a bandaid to me, we should understand the root cause and fix it.
> > > 
> > > 
> > > I think this fix is fine. This is an LTS version, so we should deliver surgical fixes. There are two possible root-causes mentioned in the cover-letter. I asked @asias to look into them. We can pursue them in parallel.
> > 
> > 
> > I am not sure it is. The full repair might lose some compacted sstables in repaired set, those sstables will be demoted to unrepaired set (since they will escape the marking as repaired by mutating repaired_at) and excluded from tombstone gc.
> 
> Right, an sstable demoted from repaired -> unrepaired will mean it will be excluded from tombstone-gc overlap checks and can cause data resurrection. So we definitely have to look into the problem of compaction allowed on repaired set in parallel to full repair.

Actuallly, if the sstable was repaired before, it will stay repaired even if it keeps the pre-repair repaired_at value. I think this is what @tarasveretilnyk talks about in the cover letter too.

**Comment by raphaelsc** [comment]
> > > > > This sounds like a bandaid to me, we should understand the root cause and fix it.
> > > > 
> > > > 
> > > > I think this fix is fine. This is an LTS version, so we should deliver surgical fixes. There are two possible root-causes mentioned in the cover-letter. I asked @asias to look into them. We can pursue them in parallel.
> > > 
> > > 
> > > I am not sure it is. The full repair might lose some compacted sstables in repaired set, those sstables will be demoted to unrepaired set (since they will escape the marking as repaired by mutating repaired_at) and excluded from tombstone gc.
> > 
> > 
> > Right, an sstable demoted from repaired -> unrepaired will mean it will be excluded from tombstone-gc overlap checks and can cause data resurrection. So we definitely have to look into the problem of compaction allowed on repaired set in parallel to full repair.
> 
> Actuallly, if the sstable was repaired before, it will stay repaired even if it keeps the pre-repair repaired_at value. I think this is what @tarasveretilnyk talks about in the cover letter too.

You're right.

**Comment by tarasveretilnyk** [comment]
> I think this problem should be fixed systematically, rather than ignoring some mutation because the sstable has already been unlinked. A sstable that has been unlinked shouldn't be a candidate for mutation of statistics. I am not a fan of checking sstable has been unlinked. We have done so in this procedure but it may happen in others too.

unlink check is small fix for 2026.1 backport. 

> A sstable in the set cannot be unlinked. A repairing sstable shouldn't be eligible for compaction, so what unlinked it before it was promoted to repaired set which happens by mutating repaired_at stat?


Case 1:
The repair ran in full mode, so it takes all sstables, also the ones that are already repaired.
The log shows the repair took this sstable:
Added sst=...3fgg02... repaired_at=4 sstables_repaired_at=4 being_repaired=5d9612b1...
repaired_at=4 and sstables_repaired_at=4 mean: this sstable was already repaired. So it was in the repaired group. The repair also set the flag being_repaired on it.
The flag did not help. Scylla first asks "is this sstable repaired?". The answer was yes, so the sstable stayed in the repaired group. Nobody looked at the flag.
During a repair, repair stops compaction only for the unrepaired group. Nothing stops compaction for the repaired group. So few milliseconds later, a normal compaction started and used this sstable:
Compacting [...3fgg02...:repaired_at=4, ...3gbb4...:repaired_at=3]
Both input sstables have repaired_at bigger than 0, which shows it is repaired group. The compaction wrote a new sstable and deleted the old ones.

Case 2:
Here the sstable was not repaired yet(repaired_at=0).
The repair master was still working, but the coordinator did not wait for it. It sent repair_update_compaction_ctrl, which removed the being_repaired flag and released the lock.
Now nothing protected the sstable and major compaction took it and deleted it


From what I can see, repairs stops compaction only for a short moment, when it makes a list of the sstables it will use. Then it starts compaction again.
After enabling, compaction can run again, so we use being_repaired for another way of protection. The code says this in a comment:
It is safe to re-enable compaction again because all sstables participating the repair have been marked as being_repaired which will be ignored by the new unrepaired compaction.
 Only compaction of the unrepaired group looks at this flag. Compaction of the repaired group does not look at it.

**Comment by tarasveretilnyk** [comment]
> ..... Compaction of the repaired group does not look at it.

There is a second problem. When repair stops compaction and takes the lock only for one group:

```
        auto& view = cg->view_for_unrepaired_data();
        auto cre = co_await db.get_compaction_manager().await_and_disable_compaction(view);
        auto lock_holder = co_await db.get_compaction_manager().get_incremental_repair_write_lock(view, "row_level_repair");

```
view_for_unrepaired_data() is the unrepaired group. The repaired group is never stopped and never locked. In full mode, repair also takes sstables that are already repaired. that are in the repaired group, which has no protection:  there is no lock there and the flag being_repaired is not checked.

**Comment by fee-mendes** [comment]
not trying to block it but... since the tests I shared are _apparently_ applicable, don't we want a follow-up commit with something covering this edge case? :)

I get it shouldn't affect later releases, but the whole idea of adding tests is to prevent we don't resurrect a previously corrected problem in the future.

File: sstables/sstables.cc [hunk 1]
```diff
@@ -1345,6 +1345,10 @@ void sstable::rewrite_statistics() {
     sstlog.debug("Rewriting statistics component of sstable {}", get_filename());
 
     auto lock = get_units(_mutate_sem, 1).get();
+    if (_unlinked) {
+        sstlog.warn("Skipping statistics rewrite of unlinked sstable {}", get_filename());
```
Comment (bhalevy) [inline sstables/sstables.cc]: Why is it a warning? If this can happen under normal circumstances then info or even debug level should be enough.

**Comment by bhalevy** [comment]
> not trying to block it but... since the tests I shared are _apparently_ applicable, don't we want a follow-up commit with something covering this edge case? :)
> 
> I get it shouldn't affect later releases, but the whole idea of adding tests is to prevent we don't resurrect a previously corrected problem in the future.

I agree we should have reproducer test(s) for this fix.

File: sstables/sstables.cc (hunk 1 above)
Comment (tarasveretilnyk) [reply sstables/sstables.cc]: Done.

**Comment by tarasveretilnyk** [comment]
> > not trying to block it but... since the tests I shared are _apparently_ applicable, don't we want a follow-up commit with something covering this edge case? :)
> > I get it shouldn't affect later releases, but the whole idea of adding tests is to prevent we don't resurrect a previously corrected problem in the future.
> 
> I agree we should have reproducer test(s) for this fix.

Added 2 reproducers.

**Review by bhalevy — APPROVED** [review]

**Comment by raphaelsc** [comment]
@asias I think we should document why it's safe to allow concurrent compaction on repaired set while it's being repaired in full mode. SSTables in repaired set are marked as repairing, but it still belongs to repaired set (not repairing), which is why a compaction can pick them. Only sstables in repairing set aren't eligible for compaction. See how table::get_compaction_reenablers_and_lock_holders_for_repair() only adds protection to unrepaired set.

**Comment by asias** [comment]
> @asias I think we should document why it's safe to allow concurrent compaction on repaired set while it's being repaired in full mode. SSTables in repaired set are marked as repairing, but it still belongs to repaired set (not repairing), which is why a compaction can pick them. Only sstables in repairing set aren't eligible for compaction. See how table::get_compaction_reenablers_and_lock_holders_for_repair() only adds protection to unrepaired set.

When a repaired sstable being repaired in full mode, it does not change the fact the sstable has been repaired. Normal  compaction could compact it with other repaired sstables without mixing repaired / repaired sstables incorrectly. After repair is done, the repaired sstable will be marked with a new repaired_at number and might be deleted  because of the compaction, it is till fine because we will use the original repaired_at number which compaction writes.

**Review by asias — APPROVED** [review]

**Comment by tarasveretilnyk** [comment]
@denesb could you please merge it?

**Comment by raphaelsc** [comment]
@tarasveretilnyk master should be fine right? since we revamped the area.

**Comment by tarasveretilnyk** [comment]
> @tarasveretilnyk master should be fine right? since we revamped the area.

Yes, I also added test reproducer PR for master https://github.com/scylladb/scylladb/pull/31789
